File: Rust/src/challenges/day_14.rs
use super::challenge::{Challenge, ChallengeErr};
use std::collections::HashMap;

enum Command<'a> {
    Mask { mask: &'a str },
    MemorySet { addr: &'a str, value: i64 },
}

impl Command<'_> {
    fn from_str(s: &str) -> Command {
        let mut parts = s.split(" = ");

        match parts.next().unwrap() {
            "mask" => Command::Mask {
                mask: parts.next().unwrap(),
            },
            val => Command::MemorySet {
                addr: &val[4..val.len() - 1],
                value: parts.next().unwrap().parse().unwrap(),
            },
        }
    }
}

struct TrieNode {
    left: Option<Box<TrieNode>>,
    right: Option<Box<TrieNode>>,
    value: i64,
}

impl TrieNode {
    fn new() -> TrieNode {
        TrieNode {
            left: None,
            right: None,
            value: 0,
        }
    }

    fn add_value_at(&mut self, address: &str, value: i64, position: usize) {
        if position >= address.chars().count() {
            self.value = value;
            return;
        }

        match address.chars().nth(position).unwrap() {
            '1' => {
                if self.right.is_none() {
                    self.right = Some(Box::new(TrieNode::new()));
                }

                self.right
                    .as_mut()
                    .unwrap()
                    .add_value_at(address, value, position + 1);
            }
            '0' => {
                if self.left.is_none() {
                    self.left = Some(Box::new(TrieNode::new()));
                }

                self.left
                    .as_mut()
                    .unwrap()
                    .add_value_at(address, value, position + 1);
            }
            'X' => {
                if self.left.is_none() {
                    self.left = Some(Box::new(TrieNode::new()));
                }
                if self.right.is_none() {
                    self.right = Some(Box::new(TrieNode::new()));
                }

                self.left
                    .as_mut()
                    .unwrap()
                    .add_value_at(address, value, position + 1);
                self.right
                    .as_mut()
                    .unwrap()
                    .add_value_at(address, value, position + 1);
            }
            _ => (),
        };

        self.value = match (self.left.as_mut(), self.right.as_mut()) {
            (Some(l), Some(r)) => l.value + r.value,
            (_, Some(r)) => r.value,
            (Some(l), _) => l.value,
            _ => value,
        };
    }
}

fn parse_program(program_text: &str) -> Vec<Command> {
    program_text.lines().map(Command::from_str).collect()
}

fn evaluate_program_v1(program: &[Command]) -> i64 {
    use Command::*;

    let mut memory: HashMap<i64, i64> = HashMap::new();
    let mut mask = "";

    for command in program {
        match command {
            Mask { mask: m } => mask = m,
            MemorySet { addr, value } => {
                let mut value = *value;
                let addr: i64 = addr.parse().unwrap();

                for (i, v) in mask.chars().rev().enumerate() {
                    match v {
                        '1' => value |= 1 << i,
                        '0' => value &= !(1 << i),
                        _ => (),
                    };
                }

                memory.insert(addr, value);
            }
        }
    }

    memory.values().sum()
}

fn apply_mask(mask: &str, addr: &str) -> String {
    let mut res = String::new();

    for (m, v) in mask.chars().zip(addr.chars()) {
        match m {
            'X' => res.push(m),
            '1' => res.push('1'),
            _ => res.push(v),
        };
    }

    res
}

fn evaluate_program_v2(program: &[Command]) -> i64 {
    use Command::*;

    let mut memory = TrieNode::new();
    let mut mask = "";

    for command in program {
        match command {
            Mask { mask: m } => mask = m,
            MemorySet { addr, value } => {
                let addr = addr.parse::<i64>().unwrap();
                let address = apply_mask(&mask, &format!("{:0>36b}", addr));

                memory.add_value_at(&address, *value, 0);
            }
        }
    }

    memory.value
}

fn find_memory_values_sum_v1(program: &[Command]) -> i64 {
    evaluate_program_v1(program)
}

fn find_memory_values_sum_v2(program: &[Command]) -> i64 {
    evaluate_program_v2(program)
}

pub struct Solution {}

impl Challenge for Solution {
    fn first_part(&self, program_text: &str) -> Result<String, ChallengeErr> {
        let program = parse_program(program_text);

        Ok(find_memory_values_sum_v1(&program).to_string())
    }

    fn second_part(&self, program_text: &str) -> Result<String, ChallengeErr> {
        let program = parse_program(program_text);

        Ok(find_memory_values_sum_v2(&program).to_string())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_memory_sum_v1() {
        let test_program_1_text = "mask = XXXXXXXXXXXXXXXXXXXXXXXXXXXXX1XXXX0X
mem[8] = 11
mem[7] = 101
mem[8] = 0";

        let test_program_1 = parse_program(&test_program_1_text);
        assert_eq!(find_memory_values_sum_v1(&test_program_1), 165);
    }

    #[test]
    fn test_memory_sum_v2() {
        let test_program_2_text = "mask = 000000000000000000000000000000X1001X
mem[42] = 100
mask = 00000000000000000000000000000000X0XX
mem[26] = 1";

        let test_program_2 = parse_program(&test_program_2_text);
        assert_eq!(find_memory_values_sum_v2(&test_program_2), 208);
    }
}

File: Python3/aoc2020/solutions/day_14.py
from __future__ import annotations

from abc import ABC
from collections import defaultdict
from typing import Dict, List, Tuple, Optional

from .solution import Solution

ADDRESS_SIZE = 36


class Command(ABC):
    pass


class Mask(Command):
    __slots__ = "value"

    INITIAL_MASK = "X" * ADDRESS_SIZE

    def __init__(self, value: str):
        self.value = value


class MemorySet(Command):
    __slots__ = "address", "value"

    def __init__(self, address: str, value: int):
        self.address = address
        self.value = value
        self.address_val = int(address, base=10)
        self.bin_address_val = bin(self.address_val)[2:].zfill(ADDRESS_SIZE)


class TrieNode:
    __slots__ = "left", "right", "value"

    left: Optional[TrieNode]
    right: Optional[TrieNode]
    value: int

    def __init__(self, value: int = 0, left: Optional[TrieNode] = None, right: Optional[TrieNode] = None):
        self.value = value
        self.left = left
        self.right = right

    def clone(self) -> TrieNode:
        if self.left and (self.left is self.right):
            cloned_node = cloned_node = self.left.clone()

            return TrieNode(self.value, cloned_node, cloned_node)

        left_copy, right_copy = None, None

        if self.left:
            left_copy = self.left.clone()
        if self.right:
            right_copy = self.right.clone()

        return TrieNode(self.value, left_copy, right_copy)

    def add_value_at(self, address: str, value: int, *, position: int = 0):
        if position >= len(address):
            self.value = value

            return

        addr_bit = address[position]
        if addr_bit == "1":
            if not self.right:
                self.right = TrieNode()
            elif self.left is self.right:
                self.right = self.left.clone()

            self.right.add_value_at(address, value, position=position + 1)
        elif addr_bit == "0":
            if not self.left:
                self.left = TrieNode()
            elif self.left is self.right:
                self.left = self.right.clone()

            self.left.add_value_at(address, value, position=position + 1)
        elif addr_bit == "X":
            node = TrieNode()
            if not self.left:
                self.left = node
            if not self.right:
                self.right = node

            self.left.add_value_at(address, value, position=position + 1)
            if self.left is not self.right:
                self.right.add_value_at(address, value, position=position + 1)

        self.value = ((self.left and self.left.value) or 0) + \
            ((self.right and self.right.value) or 0)


def parse_command(command_text: str) -> Command:
    parts = command_text.split(" = ")
    if parts[0] == "mask":
        return Mask(parts[1])

    return MemorySet(parts[0][4:-1], int(parts[1], base=10))


def parse_program(program_text: str) -> List[Command]:
    return [parse_command(c) for c in program_text.splitlines()]


def emulate_program_v1(program: List[Command]) -> int:
    memory: Dict[int, int] = defaultdict(int)
    mask = Mask.INITIAL_MASK

    for command in program:
        if isinstance(command, Mask):
            mask = command.value
        elif isinstance(command, MemorySet):
            value = command.value
            for i, v in enumerate(reversed(mask)):
                if v == "0":
                    value &= ~(1 << i)
                elif v == "1":
                    value |= (1 << i)
            memory[command.address_val] = value

    return sum(memory.values())


def emulate_program_v2(program: List[Command]) -> int:
    memory = TrieNode()
    mask = Mask.INITIAL_MASK

    def apply_mask(val: str) -> str:
        res = []

        for m, v in zip(reversed(mask), reversed(val)):
            if m == "X":
                res.append(m)
            elif m == "1":
                res.append("1")
            elif m == "0":
                res.append(v)

        return "".join(reversed(res))

    for command in program:
        if isinstance(command, Mask):
            mask = command.value
        elif isinstance(command, MemorySet):
            address = apply_mask(command.bin_address_val)
            value = command.value
            memory.add_value_at(address, value)

    return memory.value


def find_memory_values_sum_v1(program: List[Command]) -> int:
    return emulate_program_v1(program)


def find_memory_values_sum_v2(program: List[Command]) -> int:
    return emulate_program_v2(program)


class Day14(Solution):

    def first_task(self, program_text: str) -> str:
        program = parse_program(program_text)

        return str(find_memory_values_sum_v1(program))

    def second_task(self, program_text: str) -> str:
        program = parse_program(program_text)

        return str(find_memory_values_sum_v2(program))
